src/linear_algebra/sesquilinear_form.lean

```diff
@@ -363,6 +363,162 @@ lemma is_compl_span_singleton_orthogonal {B : V →ₗ[K] V →ₗ[K] K}
 
 end orthogonal
 
+/-! ### Adjoint pairs -/
+
+section adjoint_pair
+
+section add_comm_monoid
+
+variables [comm_semiring R]
+variables [add_comm_monoid M] [module R M]
+variables [add_comm_monoid M₁] [module R M₁]
+variables [add_comm_monoid M₂] [module R M₂]
+variables (B F : M →ₗ[R] M →ₗ[R] R) (B' : M₁ →ₗ[R] M₁ →ₗ[R] R) (B'' : M₂ →ₗ[R] M₂ →ₗ[R] R)
+variables (f f' : M →ₗ[R] M₁) (g g' : M₁ →ₗ[R] M)
+
+/-- Given a pair of modules equipped with bilinear forms, this is the condition for a pair of
+maps between them to be mutually adjoint. -/
+def is_adjoint_pair := ∀ ⦃x y⦄, B' (f x) y = B x (g y)
+
+variables {B B' f f' g g'}
+
+protected lemma is_adjoint_pair.eq (h : is_adjoint_pair B B' f g) :
+  ∀ {x y}, B' (f x) y = B x (g y) := h
+
+lemma is_adjoint_pair_iff_comp_eq_compl₂ :
+  is_adjoint_pair B B' f g ↔ B'.comp f = B.compl₂ g :=
+begin
+  split; intros h,
+  { ext x y, rw [comp_apply, compl₂_apply], exact h.eq },
+  { intros x y, rw [←compl₂_apply, ←comp_apply, h] },
+end
+
+lemma is_adjoint_pair_zero : is_adjoint_pair B B' 0 0 :=
+λ _ _, by simp only [zero_apply, map_zero]
+
+lemma is_adjoint_pair_id : is_adjoint_pair B B 1 1 := λ x y, rfl
+
+lemma is_adjoint_pair.add (h : is_adjoint_pair B B' f g) (h' : is_adjoint_pair B B' f' g') :
+  is_adjoint_pair B B' (f + f') (g + g') :=
+λ x _, by rw [f.add_apply, g.add_apply, B'.map_add₂, (B x).map_add, h, h']
+
+lemma is_adjoint_pair.comp {f' : M₁ →ₗ[R] M₂} {g' : M₂ →ₗ[R] M₁}
+  (h : is_adjoint_pair B B' f g) (h' : is_adjoint_pair B' B'' f' g') :
+  is_adjoint_pair B B'' (f'.comp f) (g.comp g') :=
+λ x y, by rw [linear_map.comp_apply, linear_map.comp_apply, h', h]
+
+lemma is_adjoint_pair.mul
+  {f g f' g' : module.End R M} (h : is_adjoint_pair B B f g) (h' : is_adjoint_pair B B f' g') :
+  is_adjoint_pair B B (f * f') (g' * g) :=
+λ x y, by rw [linear_map.mul_apply, linear_map.mul_apply, h, h']
```

Comment:
```suggestion
h.mul h'
```

Reply:
Gives the error:
`
type mismatch at application
  is_adjoint_pair.comp h'
term
  h'
has type
  B.is_adjoint_pair B f' g' : Prop
but is expected to have type
  ?m_3 →ₗ[?m_1] ?m_3 →ₗ[?m_1] ?m_1 : Type (max ? ?)
Additional information:
/media/moritz/ExterneCrypt/coding/lean/mathlib_mcdoll/sesquilinear_refactor/src/linear_algebra/sesquilinear_form.lean:413:1: context: switched to simple application elaboration procedure because failed to use expected type to elaborate it, error message
  type mismatch, term
    is_adjoint_pair.comp ?m_4 h
  has type
    B.is_adjoint_pair ?m_4 ?m_5 ?m_6 → B.is_adjoint_pair ?m_4 (?m_5.comp f) (comp g ?m_6)
  but is expected to have type
    B.is_adjoint_pair B (f * f') (g' * g)
`

Reply:
That means some of your arguments are explicit but should be explicit. Either way, this lemma has a trivial proof in terms of the previous one